tower-balance/src/lib.rs

```diff
@@ -1,21 +1,48 @@
+#[macro_use]
 extern crate futures;
+extern crate ordermap;
+extern crate rand;
 extern crate tower;
 extern crate tower_discover;
-extern crate ordermap;
 
 use futures::{Future, Poll, Async};
-
+use ordermap::OrderMap;
+use rand::Rng;
+use std::marker::PhantomData;
 use tower::Service;
 use tower_discover::Discover;
 
-use ordermap::OrderMap;
+pub mod load;
+pub mod select;
 
-/// Balances requests across a set of inner services using a round-robin
-/// strategy.
-pub struct Balance<T>
-where T: Discover,
+pub use load::{Load, PollLoad, WithLoad};
+pub use select::Select;
+
+/// Balances requests across a set of inner services.
+pub struct Balance<D, L, S>
+where
+    D: Discover,
+    D::Key: Clone,
+    L: WithLoad<Service = D::Service>,
+    S: Select<Key = D::Key, Service = L::PollLoad>,
 {
-    balancer: RoundRobin<T>,
+    /// Provides endpoints from service discovery.
+    discover: D,
+
+    /// Wraps endpoint services with a load metric.
+    with_load: L,
+
+    /// Determines which endpoint is ready to be used next.
+    select: S,
+
+    /// Holds the key of an endpoint that has been selected to be used next.
+    ready_endpoint: Option<D::Key>,
+
+    /// Holds all possibly-available endpoints (i.e. from `discover`).
+    endpoints: OrderMap<D::Key, L::PollLoad>,
+
+    /// Newly-added endpoints that have not yet become ready.
+    new_endpoints: OrderMap<D::Key, L::PollLoad>,
```

Comment:
Could you explain why you opted to split out `new_endpoints`?

Also, if there are two maps, why not have it be `ready_endpoints` vs. `not_ready_endpoints`, i.e., remove endpoints from `endpoints` when `poll_ready` returns `NotReady`?

Reply:
I opted to split out new_endpoints because, once we get an update from service discovery, we basically _know_ the endpoint isn't ready, so if we offer it up to the selector, it has a very low likelihood of being selected.

However, after this, the balancer is no longer (and should not be) responsible for polling all nodes for readiness (let's assume we'll end up with 100s of nodes in a LB)-- we want the selector to be responsible for this polling so it's polling the fewest nodes needed to make a balancing determination.

I think we can probably remove `new_endpoints`, as there isn't a strong motivation for it.